HACKING.md

```diff
@@ -0,0 +1,388 @@
+
+## Environment Setup
+
+To hack on Salt or the docs you're going to need to setup your development
+environment. If you already have a workflow that you're comfortable with, you
+can use that, but otherwise this is an opinionated guide for setting up your
+dev environment. Follow these steps and you'll end out with a functioning dev
+environment and be able to submit your first PR.
+
+This guide assumes at least a passing familiarity with
+[Git](https://git-scm.com/), and that you already have it installed and
+configured.
+
+## `pyenv`, Virtual Environments, and You
+
+We recommend [pyenv](https://github.com/pyenv/pyenv), since it allows
+installing multiple different Python versions, which is important for testing
+Salt across all the versions of Python that we support.
+
+### On Linux
+
+Install pyenv: 
+
+    git clone https://github.com/pyenv/pyenv.git ~/.pyenv
+    git clone https://github.com/pyenv/pyenv-virtualenv.git $(pyenv root)/plugins/pyenv-virtualenv
+
+### On Mac
+
+Install pyenv using brew:
+
+    brew update
+    brew install pyenv
+    brew install pyenv-virtualenv
+
+---
+
+Now add pyenv to your `.bashrc`:
+
+    export PATH="/root/.pyenv/bin:$PATH" >> ~/.bashrc
+    eval "$(pyenv init -)" >> ~/.bashrc
+    eval "$(pyenv virtualenv-init -)" >> ~/.bashrc
+
+For other shells, see [the pyenv instructions](https://github.com/pyenv/pyenv#basic-github-checkout).
+
+Go ahead and restart your shell. Now you should be able to install a new
+version of Python:
+
+    pyenv install 3.7.0
+
+If that fails, don't panic! You're probably just missing some build
+dependencies. Check out [pyenv common build
+problems](https://github.com/pyenv/pyenv/wiki/Common-build-problems).
+
+Now that you've got your version of Python installed, you can create a new
+virtual environment with this command:
+
+    pyenv virtualenv 3.7.0 salt
+
+Then activate it:
+
+    pyenv activate salt
+
+Sweet! Now you're ready to clone salt so you can start hacking away!
+
+### Cloning Salt
+
+Clones are so shallow. Well, this one is anyway:
+
+    git clone --depth=1 --origin salt https://github.com/saltstack/salt.git
+
+This creates a shallow clone of salt, which should be fast. Most of the time
+that's all you'll need, and you can start building out other commits as you go.
+If you *really* want all 108,300+ commits you can just run `git fetch
+--unshallow`. Then go make a sandwich because it's gonna be a while.
+
+### Fork Salt
+
+You're also going to want to head over to GitHub and create your own [fork of
+salt](https://github.com/saltstack/salt/fork). Once you've got that setup you
+can add it as a remote:
+
+    git remote add yourname <YOUR SALT REMOTE>
+
+If you use your name to refer to your fork, and `salt` to refer to the official
+Salt repo you'll never get `upstream` or `origin` confused.
+
+### Install sphinx
+
+In order to build the docs, you'll need to install sphinx into your virtual
+environment:
+
+    python -m pip install sphinx
```

Comment:
`nox -e docs` or `nox -e docs-html` or `nox -e docs-man`

Reply:
Ie, don't install sphinx directly...

Reply:
Or use `nox -e 'docs-html(compress=False, clean=True)'` (`nox` commands can be viewed with `nox -l`)

Reply:
Yes, that :)

Reply:
Added that.